[libpas] Improve PGM metadata management

zakariaridouh opened the pull request:
#### f4ebee8774596e8435dcba922f8ff30579469060
<pre>
[libpas] Improve PGM metadata management
<a href="https://bugs.webkit.org/show_bug.cgi?id=285557">https://bugs.webkit.org/show_bug.cgi?id=285557</a>
<a href="https://rdar.apple.com/142506754">rdar://142506754</a>

Reviewed by Brandon Stewart.

Simplify the metadata management for PGM in libpas. Use a circular
buffer to store metadata upon deallocations.

Add tests to verify improvement.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_check_exists):
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_full): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_empty): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_add): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_remove): Deleted.
(pas_probabilistic_guard_malloc_manage_metadata): Deleted.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMMetaData):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(addPGMTests):

Canonical link: <a href="https://commits.webkit.org/288785@main">https://commits.webkit.org/288785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bcb51e32bfe285e3d0c598b79f28ec8eb4be593

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84294 "26 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3916 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38599 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89372 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35303 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11892 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65558 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23395 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87340 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3003 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76591 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45858 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2954 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30821 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34352 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77258 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73854 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90752 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83311 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11561 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8397 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html media/video-replaces-poster.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74012 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11788 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72416 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73213 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18137 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17536 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15978 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2926 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11513 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105730 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11362 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25525 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13135 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -79,7 +73,7 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
 {
     pas_heap_lock_assert_held();
-    static const bool verbose = false;
+    static const bool verbose = true;
```
Should be `false` 😊

zakariaridouh replied:
Yup, leftover from some testing.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,65 @@ void testPGMErrors() {
+    pas_root* root = pas_root_create();
+    pas_heap_lock_unlock();
+
+    int* int_arr1 = static_cast<int*>(bmalloc_allocate(30000 * sizeof(int), pas_non_compact_allocation_mode));
```
Just a thought for making less redundant code, you could just store the results in an array of `int*` and then loop over the array when freeing.

stwrt replied:
And since the allocations are all the same size we could simplify this with.

```cpp
for (int i = 0; i < 12; i++) {
    arr[i] = static_cast<int*>(bmalloc_allocate(30000 * sizeof(int), pas_non_compact_allocation_mode));
    CHECK(arr[i]);
}
```

zakariaridouh replied:
Fixed =)

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -208,7 +202,15 @@ void pas_probabilistic_guard_malloc_deallocate(void* mem)
     value->free_status = true;
-    pas_probabilistic_guard_malloc_manage_metadata(entry);
+    pas_ptr_hash_map_entry* old_entry = pgm_metadata_vector[pgm_metadata_index];
+    if (old_entry != NULL && entry && entry->value) {
```
We can just write `old_entry` instead.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,73 @@ void testPGMErrors() {
+    size_t count = 0;
+    while (metadata_array[count] != NULL)
+        count++;
+    CHECK(count <= MAX_PGM_DEALLOCATED_METADATA_ENTRIES);
```
I think we are expecting the count to always be equal to MAX_PGM_DEALLOCATED_METADATA_ENTRIES.

stwrt reviewed (commented):
Overall I think this is looking pretty good. Just a few minor points.
Can you confirm that the CIs succesfully passed locally and no regressions were found in benchmarking?

stwrt commented on Source/bmalloc/libpas/libpas.xcodeproj/xcshareddata/xcschemes/test_pas.xcscheme:
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This should be added in a separate PR.

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -208,7 +202,15 @@ void pas_probabilistic_guard_malloc_deallocate(void* mem)
     value->free_status = true;
-    pas_probabilistic_guard_malloc_manage_metadata(entry);
+    pas_ptr_hash_map_entry* old_entry = pgm_metadata_vector[pgm_metadata_index];
+    if (old_entry && entry && entry->value) {
```
I was looking above and we already have a check for `entry`.

```
if (!entry || !entry->value)
        return;
```

stwrt commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -208,7 +202,15 @@ void pas_probabilistic_guard_malloc_deallocate(void* mem)
-    pas_probabilistic_guard_malloc_manage_metadata(entry);
+    pas_ptr_hash_map_entry* old_entry = pgm_metadata_vector[pgm_metadata_index];
+    if (old_entry && entry && entry->value) {
+        pas_pgm_storage *old_entry_value = (pas_pgm_storage *) old_entry->value;
```
`old_entry_value` is only ever used once; do we need create a new variable for it? I don't think the variable helps with clarifying what it's purpose is either.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,115 @@ void testPGMErrors() {
+    pas_ptr_hash_map* hash_map = root->pas_pgm_hash_map_instance;
+    CHECK(hash_map);
+
+    CHECK_EQUAL(hash_map->key_count, 11);
```
If `MAX_PGM_DEALLOCATED_METADATA_ENTRIES` ever changes this test may need to be updated here.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -162,7 +165,7 @@ void testPGMMetaData()
     pas_ptr_hash_map_entry** metadata_array = pas_probabilistic_guard_malloc_get_metadata_array();
 
-    size_t num_allocations = 1000, num_success_allocs = 0, num_de_allocs = 0;
+    size_t num_allocations = 80, num_success_allocs = 0, num_de_allocs = 0;
```
Don't quite understand this change.

zakariaridouh replied:
Leftover from some other testing.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,117 @@ void testPGMErrors() {
+    }
+
+    pas_heap_lock_lock();
+    // Deallocate all previous allocations
```
This comment isn't quite accurate. It deallocates all previous allocations except the one on line 264.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,117 @@ void testPGMErrors() {
+    }
+
+    pas_heap_lock_lock();
+    // Deallocate all previous allocations
```
Not accurate.

stwrt commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -252,6 +253,117 @@ void testPGMErrors() {
+    }
+
+    pas_heap_lock_lock();
+    // Deallocate all previous allocations
```
same comment as above.

stwrt reviewed (approved):
This PR great simplifies the deallocation metadata handling and adds some good test cases to catch a few bugs.

I think there is a couple of minor points around the test cases that need to be addressed.

As long as the libpas tests pass and benchmarks show now regressions I think this should be good.

zakariaridouh committed:
[libpas] Improve PGM metadata management
https://bugs.webkit.org/show_bug.cgi?id=285557
rdar://142506754

Reviewed by Brandon Stewart.

Simplify the metadata management for PGM in libpas. Use a circular
buffer to store metadata upon deallocations.

Add tests to verify improvement.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_check_exists):
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_full): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_empty): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_add): Deleted.
(pas_probabilistic_guard_malloc_pgm_metadata_buffer_remove): Deleted.
(pas_probabilistic_guard_malloc_manage_metadata): Deleted.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMMetaData):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(addPGMTests):

Canonical link: https://commits.webkit.org/288785@main
